packages/test/src/util/browser-tests-runner/browser-dependencies/index.js

```diff
@@ -0,0 +1,141 @@
+// Capture client logs and forward to server.
+["log", "info", "error"].forEach(method => {
+  const fn = console[method] || console.log || (() => {});
+  console[method] = (...args) => {
+    send(["console", method, args]);
+    fn.apply(console, args);
+  };
+});
+
+// Forward uncaught excpections.
+window.addEventListener("error", ({ error }) => console.error(error));
+
+require("chai").config.includeStack = true;
+const BrowserContext = require("./browser-context");
+let buffer = [];
+let options;
+
+try {
+  options = JSON.parse(decodeURIComponent(window.location.search.slice(1)));
+} catch (_) {
+  options = {};
+}
+
+// Apply mocha options.
+const mochaOptions = {
+  ...options.mochaOptions,
+  reporter: "spec",
+  useColors: true,
+  ui: "bdd"
+};
+
+Object.keys(mochaOptions).forEach(key => {
+  if (typeof mocha[key] === "function") {
+    mocha[key](mochaOptions[key]);
+  }
+});
+
+window.__run_tests__ = done => {
+  const socket = require("engine.io-client")(`ws://${location.host}`);
+  socket.once("open", () => {
+    const runner = mocha.run();
+    const fails = [];
+    let timeout;
+    queueFlush(); // We write the buffered console output on an interval.
+
+    runner.on("fail", (test, err) => {
+      fails.push({
+        name: test.title,
+        result: false,
+        message: err.message,
+        stack: err.stack,
+        titles: flattenTitles(test)
+      });
+    });
+
+    runner.on("end", () => {
+      window.mochaResults = runner.stats;
+      window.mochaResults.reports = fails;
+      clearTimeout(timeout);
+      flush(() =>
+        done({
+          success: !fails.length,
+          coverage: window.__coverage__
+        })
+      );
+    });
+
+    function flush(cb) {
+      if (buffer.length) {
+        socket.send(JSON.stringify(buffer), cb);
+        buffer = [];
+      }
+    }
+
+    function queueFlush() {
+      timeout = setTimeout(() => flush(queueFlush), 100);
+    }
+  });
+};
+
+window.__init_test__ = (test, component, func) => {
+  test.component = component;
+  const context = new BrowserContext(test);
+  window.test = (name, handler) => runTest(it, name, handler, context);
+  Object.keys(it).forEach(function(key) {
+    if (typeof it[key] === "function") {
+      window.test[key] = (name, handler) =>
+        runTest(it[key], name, handler, context);
+    }
+  });
+
+  let desc = test.componentName;
+
+  if (test.groupName) {
+    desc += " - " + test.groupName;
+  }
+
+  describe(desc, func);
+
+  window.test = null;
+};
+
+function runTest(it, name, handler, context) {
+  if (handler.length <= 1) {
+    it(name, function() {
+      context.name = name;
+      const testFunction = handler.call(this, context);
+      if (isPromise(testFunction)) {
+        testFunction.then(() => context._afterTest());
```

Comment:
This looks like it is going to result in `context._afterTest()` never getting called if the test throws an error. I think we should change this to catch all errors, call `context._afterTest()` and then rethrow the error (e.g. behave like a `finally`). Something like this:

```js
if (isPromise(testFunction)) {
  return testFunction
    .then(() => context._afterTest())
    .catch((err) => {
      context._afterTest();
      throw err;
    });
}
```

Reply:
That portion was c&p from before but I’ll make sure it’s working properly when I get a sec. Thanks for pointing it out!